### DISCUSSION TITLE
Skipping generating static sites at build time (instead: only at runtime) [getStaticProps no data in build phase]

### DISCUSSION THREAD

**Discussion by bennettdams** [vercel/next.js#28298] [Help]
I want to use the same application for different customers, where there's a different database for each (which could be on-premise).
As a result, I have no data at the build phase (e.g. in CI/CD) that I could use to create static sites.

I thought about skipping generating the sites in `getStaticProps` via an environment variable.
When building the sites, I could tell Next.js to not use any data, something like this:

```tsx
export const getStaticProps: GetStaticProps<HomePageProps> = async () => {
  const isBuildPhase = process.env.IS_BUILD;

  const data = isBuildPhase ? null : await fetchData();

  return {
    props: {
      data: data ?? null,
    },
    revalidate: 5 * 60,
  };
};
```

Now, I want to create the sites only at runtime (with `next start`), because at runtime, the built application has access to its database. In each `getStaticProps` the environment variable will be configured so data is fetched. When the application starts initially, it will generate all static sites when they're accessed.

Are there big downsides with this approach?
Are there maybe better solutions to this problem?

---

Update: I created a feature request (including another topic) to show why this could be beneficial:

[Feature request: Same Next.js application for different customers/databases and protected static sites](https://github.com/vercel/next.js/discussions/28959)

**Comment by tilvan** [comment]
I have similar question.

I use `getInitialProps` in `_app.ts`. Where I fetch API for initial data.
And my question is how I can skip API fetching on build.
I want to fetch API only on http request.

[Here is full example](https://github.com/vercel/next.js/discussions/28362)

**Comment by hampusborgos** [comment]
This will not work, as NextJS will serve the page with `data: null` on the first incoming request (or after the `revalidate` timeout expires). This means some of your users will just see a blank screen (or whatever you render when `props.data` is null and need to refresh.

You could hack around this, by checking client-side if `props.data === null` and issue `window.location.reload()` until Next has fetched the props. But it is still not a good user-experience.

Ideally an option like:
```
return {
  revalidate: 5 * 60,
  buildTime: false
}
```
should be added to enable this use-case.
